### DISCUSSION TITLE
Allow to get property type as script-instance in GDScript

### DISCUSSION THREAD

**Discussion by zaevi** [godotengine/godot-proposals#5305] [Scripting]
For example:
```gdscript
tool
extends EditorScript

class SubClassA extends Reference:
    pass

const SubClassB = preload("SubClassB.gd")

# SubClassA and SubClassB are GDScript instances and we can declare typed variable
var instanceA : SubClassA
var instanceB : SubClassB
```

I want to know the actual type of property at runtime, but currently we can only know that they are `TYPE_OBJECT` via `get_property_list`. 

So I hope GDScript exposes a method that can get a script instance of property type:

```gdscript
func _run():
    var script = self.get_script() as GDScript
    
    var typeA = script.get_property_type("instanceA")
    assert(typeA == SubClassA)

    var typeB = script.get_property_type("instanceB")
    instanceB = typeB.new()
```

If implemented, we can do cool things like dependency-injection for properties.

**Comment by dalexeev** [comment]
You can already instantiate scripts dynamically. And even the `is` operator works correctly.

```gdscript
var script = load('res://my_script.gd')
var script_1 = load('res://my_script_1.gd')
var instance = script.new()
print(instance is script)
print(instance is script_1)
print(instance is Node)
print(instance is Reference)
```

**Reply by zaevi** [reply]
No I want to get properties' type as GDScript instance. It's similar to reflection in C#:
```gdscript
const MyScript = preload("my_script.gd")

# declare a property member in class:
var instance : MyScript

# reflect the property's info
var property_info = get_script().get_property_info()
var property_type = property_info.get_type() as GDScript
assert(property_type == MyScript)
```

**Reply by dalexeev** [reply]
```gdscript
object.property.get_script()
```

**Reply by zaevi** [reply]
You cannot use like this if the property is null. What I need is something like reflection in C#.

**Reply by dalexeev** [reply]
So check this: 1. `typeof` / `is Object`, 2. `get_script() != null`.

Yes, different types are not unified now, there are `typeof`, `get_class` and `get_script`. But this is a slightly different issue that is already being discussed (godotengine/godot#21789).
